### PR TITLE
[Impeller] Make paths externally immutable, update all tests to use PathBuilder to create Path.

### PR DESCRIPTION
**Pull request by jonahwilliams**
This is a design problem in the Flutter API, that in general you can't assume a Path hasn't changed from the last time you used it. We're so close, so just move the mutation methods to PathBuilder :)

Fixes https://github.com/flutter/flutter/issues/133880

### CODE REVIEW

File: impeller/geometry/path.h [hunk 1]
```diff
@@ -48,8 +48,8 @@ enum class Convexity {
 ///             All shapes supported by Impeller are paths either directly or
 ///             via approximation (in the case of circles).
 ///
-///             Creating paths that describe complex shapes is usually done by a
-///             path builder.
+///             Paths are externally immutable once created, Creating paths must
```
Comment (chinmaygarde) [inline impeller/geometry/path.h:51]: 👍🏽

**Review by chinmaygarde — APPROVED** [review]

**Comment by jonahwilliams** [comment]
So we might have a bug in path builder, or I'm misunderstanding the point of the test. We had a test case "PathCreatePolyLineDoesNotDuplicatePoints" 

https://github.com/flutter/engine/blob/f15128970f99e2d2016e63dd4ad6b7cf10a00bb6/impeller/geometry/geometry_unittests.cc#L2009-L2026

I updated it to use the PathBuilder API in a way that I think is correct. However, now it seems like the polyline does in fact have duplicated points.

**Comment by jonahwilliams** [comment]
The points I get from the tests are now:

```
[ERROR:flutter/impeller/geometry/geometry_unittests.cc(2035)] (10, 10)
[ERROR:flutter/impeller/geometry/geometry_unittests.cc(2035)] (20, 20)
[ERROR:flutter/impeller/geometry/geometry_unittests.cc(2035)] (20, 20)
[ERROR:flutter/impeller/geometry/geometry_unittests.cc(2035)] (30, 30)
[ERROR:flutter/impeller/geometry/geometry_unittests.cc(2035)] (40, 40)
[ERROR:flutter/impeller/geometry/geometry_unittests.cc(2035)] (50, 50)
```

WHich I think is because PathBuilder is inserting Moves. However, we're only ever building paths with PathBuilder today, so I'm not sure if the expectation is overly specific to the original bug.

FYI @bdero

File: impeller/geometry/geometry_unittests.cc [hunk 2]
```diff
@@ -2007,14 +2020,14 @@ TEST(GeometryTest, CubicPathComponentPolylineDoesNotIncludePointOne) {
...
-  Path path;
-  path.AddContourComponent({10, 10});
-  path.AddLinearComponent({10, 10}, {20, 20});
-  path.AddLinearComponent({20, 20}, {30, 30});
-  path.AddContourComponent({40, 40});
-  path.AddLinearComponent({40, 40}, {50, 50});
+  PathBuilder builder;
+  builder.MoveTo({10, 10});
+  builder.AddLine({10, 10}, {20, 20});
+  builder.AddLine({20, 20}, {30, 30});
```
Comment (bdero) [inline impeller/geometry/geometry_unittests.cc:2026]: `Add*` inserts a new contour -- this actually needs to be `LineTo` to be equivalent.

```cpp
builder.MoveTo({10, 10});
builder.AddLine({20, 20});
builder.AddLine({30, 30});
```

And no, you're not crazy. Yes, the linear component has redundant data.

Comment (jonahwilliams) [reply impeller/geometry/geometry_unittests.cc:2026]: Aghh, good catch. Thank you

**Pull request merged**